sdk/core/azure-core/src/http/winhttp/win_http_request.hpp

```diff
@@ -0,0 +1,213 @@
+   *
+   * The WinHttpRequest object has a WinHttpAction associated it to convert asynchronous WinHTTP
+   * operations to synchronous operations.
+   *
+   */
+  class WinHttpAction final {
+    friend class WinHttpRequest;
+
+    // Containing HTTP request, used during the status operation callback.
+    WinHttpRequest* m_httpRequest{};
+    std::mutex m_actionCompleteMutex;
+    //    std::condition_variable m_actionCompleted;
+    wil::unique_event m_actionCompleteEvent;
+    DWORD m_stowedError{};
+    DWORD_PTR m_stowedErrorInformation{};
+    DWORD m_bytesAvailable{};
+
+    /*
+     * Callback from WinHTTP called after the TLS certificates are received when the caller sets
+     * expected TLS root certificates.
+     */
+    static void CALLBACK StatusCallback(
+        HINTERNET hInternet,
+        DWORD_PTR dwContext,
+        DWORD dwInternetStatus,
+        LPVOID lpvStatusInformation,
+        DWORD dwStatusInformationLength) noexcept;
+
+    /*
+     * Callback from WinHTTP called after the TLS certificates are received when the caller sets
+     * expected TLS root certificates.
+     */
+    void OnHttpStatusOperation(
+        HINTERNET hInternet,
+        DWORD internetStatus,
+        LPVOID statusInformation,
+        DWORD statusInformationLength);
+
+  public:
+    /**
+     * @brief Create a new WinHttpAction object associated with a specific WinHttpRequest.
+     *
+     * @param request Http Request associated with the action.
+     */
+    WinHttpAction(WinHttpRequest* request)
+        : m_httpRequest(request), m_actionCompleteEvent(CreateEvent(nullptr, TRUE, FALSE, nullptr))
+    {
+      if (!m_actionCompleteEvent)
+      {
+        throw std::runtime_error("Error creating Action Complete Event.");
+      }
+    }
+    /**
+     * @brief WaitForAction - Waits for an action to complete.
+     *
+     * @remarks The WaitForAction method waits until an action initiated by the `callback` function
+     * has completed. It polls with a timeout of milliseconds, checking to see if the provided
+     * context parameter has timed out.
+     *
+     * @param callback - Callback used to initiate an action. Always called in the waiting thread.
+     * @param pingDuration - The time to wait for a ping to complete.
+     * @param context - Context for the operation.
+     *
+     * @returns true if the action completed normally, false if there was an error.
+
+     * @remarks If there is an error, the caller can determine the error code by calling
+     * GetStowedError() and GetStowedErrorInformation()
+     */
+    bool WaitForAction(
+        std::function<void()> callback,
+        Azure::Core::Context const& context,
+        Azure::DateTime::duration const& pingDuration = std::chrono::milliseconds(800));
+
+    /**
+     * @brief Notify a caller that the action has completed successfully.
+     */
+    void CompleteAction();
```

Comment:
Take this suggestion with a grain of salt (since I don't fully understand the implementations and caller/callee semantics for this function), but should it be renamed as the following, given the doc comment:
```suggestion
    void ActionCompleted();
```

Reply:
`CompleteAction` is an active event - it tells the waiting caller that it's ok to proceed. It's not a passive notification (which is implied by `ActionCompleted`.